### ISSUE TITLE
Feature Request: Userfields filter

### ISSUE THREAD

**Issue by kmp8072** [michael-milette/moodle-filter_filtercodes#293]
### Prerequisites

- [X] I am requesting the new feature to the correct repository.
- [X] I checked the latest version on GitHub and the [README.md](../blob/master/README.md) documentation to ensure that this functionality does not already exist.
- [X] I searched to make sure that this feature has not already been requested.
- [X] I have, or I know someone who has, the ability and willingness to test the new feature on a Moodle LMS-based site.
- [X] I agree to submit one feature request per issue ticket (do not submit multiple requests in one issue ticket).
- [X] I understand that development and support is provided by volunteer(s) and that all final implementation decisions are up TNG Consulting Inc.

### Feature Request - The User Story

I would request for a new type of conditional fiter support, based on user fields like country or language etc. 
Problem: I want to show some content based on user field, like user timezone, country, authentication methods etc. Currently there is no direct way to do so, untill I create a cohort and show content based on cohort idnumber, but it makes us to do multiple things for a simple task of content showing.
Suppose I want to some content to users of a particular country only, then {ifuserfield_country countrycode}{/ifuserfield_country} would be helpful. 

### Planning on submitting a solution in a pull request (PR)?

Maybe

### Code of Conduct

- [X] I have read and agree to the project's <a href="../blob/master/CODE_OF_CONDUCT.md" target="_blank">code of conduct</a>.

**Comment by michael-milette** [comment]
Hi @kmp8072 ,

Can you give me the details of a actual real-life situation where you will use it?

Best regards,

Michael Milette

**Comment by Learnymag** [comment]
Hi @michael-milette, Here is an example.

I created a custom field named « client. » I created a custom report for each client.

I want to create an HTML block on the dashboard and grant access to a specific URL for each client.

Below is a simplified example of the code :
{ifuserfield_client Exemple1}<a href="https://xxxxx.fr/reportbuilder/view.php?id=20" </a>
{ifuserfield_client Exemple2}<a href="https://xxxxx.fr/reportbuilder/view.php?id=21" </a>

Best regards,

Boris

**Comment by michael-milette** [comment]
Hi @kmp8072 ,

Can you give me an actual usage instead of an abstract example? Imagine such a tag already exists, show me an example with real content for at least 3 countries.

Best regards,

Michael

**Comment by kmp8072** [comment]
Hello @michael-milette 
Here, I am putting few examples of the use cases:

**At Dashboard**
{ifuserfield_country US}A new certification program has been added. You can join  at [course/program link]{/ifuserfield_country}
{ifuserfield_client GB}We are having an event at Town Hall, you are most welcome to join us.{/ifuserfield_country}
{ifuserfield_client CA}Hello, there is a discussion going on about product launching in Canada, please add your suggestions {forumembed - using forumembed filter}{/ifuserfield_country}

**At course level**
{ifuserfield_country US}A new course as a replacement for this course has been launched and you should do the new course at [courselink]. Please don't register for this course. You won't we able to access content because there is restriction applied on the basis of user's country. Now this course is relevant for only Canadian users.[course/program link]{/ifuserfield_country}
  -- inside a page activity where as content of page i am embedding video links
{ifuserfield_country US}[USvideolink]{/ifuserfield_country}
{ifuserfield_country CA}[CAvideolink]{/ifuserfield_country}
  -- inside quiz creation where quiz is part of course completion and I can't restrict this activity based on user profile 
      otherwise course completion won't be achieved. Questions are different for different countries. **_This can be sparingly 
      useful but we have got situations like this and created a whole new course for that purpose_**.
 Question 1. 
    {ifuserfield_country CA}[CA question text]{/ifuserfield_country}
    {ifuserfield_country US}[US question text]{/ifuserfield_country}
     Option 1.
     {ifuserfield_country CA}[CA option 1 text]{/ifuserfield_country}
     {ifuserfield_country US}[US option 1 text]{/ifuserfield_country}

I hope above use cases make sense. 

Best regards,

Krishna Mohan

**Comment by michael-milette** [comment]
Hi @kmp8072 ,

Thank you for the great examples. I can now see how this would be useful.

Please be aware that filters do not work in all quiz questions. I tried to get that fixed several years ago but I was told that it did not make sense.

Thanks again for taking the time to prepare the examples. I will let you know when I have something for you to try out.

Best regards,

Michael

**Comment by michael-milette** [comment]
A new {ifprofile shortname ...}{/ifprofile} tag was recently added to FilterCodes. This new tag includes a lot of flexibility to meet your needs. Here is a list of the various options you can use with it:

- {ifprofile shortname is "value"}...{/ifprofile}
- {ifprofile shortname not "value"}...{/ifprofile}
- {ifprofile shortname is ""}...{/ifprofile}
- {ifprofile shortname not ""}...{/ifprofile}
- {ifprofile shortname contains "value"}...{/ifprofile}
- {ifprofile shortname in "value"}...{/ifprofile}

Notes: 

- The **shortname** can be any visible custom profile fields, unless the FilterCodes ifprofilefiedonlyvisible setting is unchecked, as well as the core profile fields including id, username, auth, idnumber, email, institution, department, city, country, timezone, and lang core profile fields. Note: You must compare to the value which is stored in the database, which may not be the same as what is displayed. For example, you must specify CA for the country of Canada, not Canada. If your profile field contains a multi-language string, you need to compare it to that string as it is before it is processed by Moodle language filters.
- The **"value"** should be replaced with the actual value to which the contents of the profile field.
- The **"value"** must always be enclosed in quotes. Use an empty set of quotes to compare the value to an empty field.
- The only comparison operators currently supported are 'is', 'not', 'contains' and 'in'.

Here are a few examples:

- {ifprofile timezone is "99"}Don't forget to set the time zone in your profile{/ifprofile}
- {ifprofile country is "CA"}Welcome Canadian student{/ifprofile}
- {ifprofile country not "CA"}Welcome international student{/ifprofile}
- {ifprofile country in "BS,BM,CA,US,MX,GL,PM,TC"}Welcome North American{/ifprofile} (I know the list is incomplete)
- {ifprofile city is ""}Don't forget to enter your city/town in your profile{/ifprofile}
- {ifprofile lang not "en"}Some courses may not be available in your preferred language.{/ifprofile}

This new tag is currently only available here on GitHub. If you would be so kind, I would be thankful if you would test it out and let me know if this works for you. If everything goes well, the {ifprofile shortname ...} tag will be included in the next release of FilterCodes on Moodle.org.

Let me know if you have any questions.

Best regards,

Michael Milette

**Comment by michael-milette** [comment]
Also see #84 .

**Comment by kmp8072** [comment]
Hi Michael,
Thank you so much, words cannot express my gratitude. These improvements will make this plugin more useful for whole community.

**Comment by michael-milette** [comment]
You're very welcome. I am glad you find the new tag helpful.

Feel free to star this plugin and leave me a recommendation on LinkedIn at https://www.linkedin.com/in/michaelmilette

Best regards,

Michael

**Comment by kmp8072** [comment]
Sure, I will do it happily.

Regards
Krishna Mohan Prasad.

**Comment by michael-milette** [comment]
Hi,

I just wanted to let you know that I updated FilterCodes again yesterday which fixed a bug with the new {ifprofile shortname ...} tag. While it worked great when a user was logged in, it wasn't handling situations where the user was logged out. In addition to fixing this bug, I also added documentation and examples on this tag in the [README.md](https://github.com/michael-milette/moodle-filter_filtercodes#is-there-more-information-about-the-ifprofile-shortname--tag) file.

Let me know if you come across any other issues.

Best regards,

Michael
